### PR TITLE
✨➕🚧 #186: add react-hook-form + clinician invite form

### PR DESCRIPTION
**Pull request by samrichca**
### Consent UI

- Add [react-hook-form](https://react-hook-form.com/), [react-select](https://react-select.com/home), and [zod resolver for RHF](https://github.com/react-hook-form/resolvers#zod) for validation.
- Create an unstyled draft of the clinician invite form
- Add dictionaries for clinician invite form & (shared) form labels & errors
- Add generic components for building forms
- Add new README for forms

### Types

Update types & tests for name & phone number regexes: handle empty & whitespace strings

### Special Instructions

Before running these changes, you will need to install react-hook-form, react-select, and @hookform/resolvers:
```
pnpm i
```

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [x] Manual testing completed
- [x] Builds locally without errors or warnings
- [x] Tests are updated (if required) and passing
- [x] PR feedback has been addressed **if applicable**
- [x] New environment variables added to `.env.schema` files, `README.md` - N/A
- [x] Added copyrights to new files

### CODE REVIEW

File: apps/consent-ui/src/components/Form/FormField/index.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		description,
+		error,
+		name,
+		label,
+		required,
+	};
+
+	return (
+		<DefaultFieldSet {...wrapperProps}>
+			<TextInput {...inputProps} />
```
Comment (joneubank) [inline apps/consent-ui/src/components/Form/FormField/index.tsx:81]: Will this eventually change component used based on the prop `type`?

File: apps/consent-ui/src/components/Form/DefaultFieldSet/index.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { ReactNode } from 'react';
+
+const DefaultFieldSet = ({
```
Comment (joneubank) [inline apps/consent-ui/src/components/Form/DefaultFieldSet/index.tsx:22]: I would remove `Default` from the name. Can specify names later if we have multiple implementations that need differentiation.

File: packages/types/tests/entities/Name.spec.ts [hunk 3]
```diff
@@ -30,5 +30,8 @@ describe('Name', () => {
 		expect(Name.safeParse('Homer_Simpson').success).false;
 		expect(Name.safeParse('-Homer Simpson').success).false;
 		expect(Name.safeParse('Homer Simpson1').success).false;
+		expect(Name.safeParse(undefined).success).false;
+		expect(Name.safeParse(null).success).false;
+		expect(Name.safeParse('').success).false;
```
Comment (samrichca) [inline packages/types/tests/entities/Name.spec.ts:35]: empty strings pass some zod validation, so we should add them to unit tests when we're also testing undefined & null.

File: apps/consent-ui/src/i18n/locales/en/index.ts [hunk 4]
```diff
@@ -42,6 +45,7 @@ import historyOfCancerDictionary from './history-of-cancer';
 const dictionaries = {
 	ancestry: ancestryDictionary,
 	'birth-sex': birthSexDictionary,
+	'clinician-invite-form': clinicianInviteFormDictionary,
```
Comment (justincorrigible) [inline apps/consent-ui/src/i18n/locales/en/index.ts:48]: Note: Not asking for changes.

For my own understanding, what was the reasoning for this project to deviate from the JS-idiomatic camelCase for keys/variables?

File: apps/consent-ui/src/i18n/locales/en/forms.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+const dictionary = {
+	'indicates-required-field': 'indicates required field',
+	'select-placeholder': 'Select an option...',
+} satisfies Record<string, string>;
```
Comment (joneubank) [inline apps/consent-ui/src/i18n/locales/en/forms.ts:23]: Repeat of previous comment. We want all french dictionaries to satisfy the shape of the english version of the same dictionary. This will let TS enforce that they have the same shape.

File: apps/consent-ui/src/i18n/locales/fr/form-labels.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	'first-name': 'Prénom',
+	'guardian-email': 'Addresse e-mail du gardien',
+	'guardian-name': 'Nom du gardien',
+	'guardian-phone': 'Numéro de téléphone du gardien',
+	'guardian-relationship': 'Relation du gardien avec le participant',
+	'last-name': 'Nom de famille',
+	'preferred-name': 'Nom préféré',
+	email: 'Adresse e-mail',
+	phone: 'Numéro de téléphone',
+} satisfies FormLabelsDictionary;
```
Comment (joneubank) [inline apps/consent-ui/src/i18n/locales/fr/form-labels.ts:38]: Like this! 👍

File: apps/consent-ui/src/i18n/locales/en/index.ts (hunk 4 above)
Comment (samrichca) [reply apps/consent-ui/src/i18n/locales/en/index.ts:48]: I'm not sure actually

File: apps/consent-ui/src/components/Form/FormField/index.tsx (hunk 1 above)
Comment (samrichca) [reply apps/consent-ui/src/components/Form/FormField/index.tsx]: the `type` just changes the type attribute in the HTML input

- email inputs use <TextFieldset type="email" /> - they work like a basic text input, but the `type` tells your browser to show an email keyboard (on mobile) and provide email options for autocomplete
- date inputs - <DateFieldset /> with a datepicker library
- checkbox inputs - <CheckboxFieldset />

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+	const {
+		getRecaptchaToken,
+		onRecaptchaChange,
+		recaptchaCheckboxRef,
+		recaptchaError,
+		resetRecaptcha,
+		setRecaptchaError,
+	} = useRecaptcha();
+
+	const handleRecaptchaChangeDemo = () => {
```
Comment (mistryrn) [inline apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:91]: Is this still for demo purposes only, or can we drop the `Demo` suffix from the function name here?

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+		const token = getRecaptchaToken();
+		token && setRecaptchaError('');
+		onRecaptchaChange();
+	};
+
+	// submit form
+	const [successMessageDemo, setSuccessMessageDemo] = useState('');
+	const onSubmit: SubmitHandler<ClinicianInviteFormValidation> = (data, event) => {
+		event?.preventDefault();
+		console.log('SUBMIT DATA', data);
```
Comment (mistryrn) [inline apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:101]: We'll want to clean up the logs before merging, but I assume this is just for testing/dev purposes at the moment so it's fine for now.

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+						<RequiredAsterisk /> {textDict['indicates-required-field']}
+					</p>
+					<TextFieldSet
+						error={errors.participantFirstName?.type && errorsDict['required']}
+						label={labelsDict['first-name'] || ''}
+						name="participantFirstName"
+						required
+					/>
+					<TextFieldSet
+						error={errors.participantFirstName?.type && errorsDict['required']}
```
Comment (mistryrn) [inline apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:153]: Noticed that these are all listed as `errors.participantFirstName?.type`, even for different fields like `participantLastName` – does it make a difference if they're all the same, or should they match up with their respective field `name` values?

File: apps/consent-ui/src/components/Form/README.md
Comment (mistryrn) [inline apps/consent-ui/src/components/Form/README.md:1]: Thanks for adding some documentation for this! 😄 Can we move it to the root `docs` directory to keep this kind of technical documentation in the same place?

**Review by mistryrn — COMMENTED** [review]
Haven't completed my review yet, but leaving a few comments for now before the meeting. Lots of good stuff so far!

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx (hunk 9 above)
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: fixed during demo, thanks for pointing it out!

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx (hunk 7 above)
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: removed, I'm keeping the success message as demo since it'll be replaced with a redirect to the homepage (added a comment to this effect)

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx (hunk 8 above)
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:103]: removed!

File: packages/types/src/entities/ClinicianInvite/ClinicianInviteForm.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+// this is specifically for validating the Clinician Invite form in consent UI
+// using the zod resolver in react-hook-form.
+// keep it separate from API schemas.
+
+export const ClinicianInviteFormValidation = z
+	.object({
+		clinicianFirstName: Name,
+		clinicianInstitutionalEmailAddress: z.string().email(),
+		clinicianLastName: Name,
+		clinicianTitleOrRole: z.string().trim().nonempty(),
```
Comment (alyssadsouza) [inline packages/types/src/entities/ClinicianInvite/ClinicianInviteForm.ts:38]: I was looking at the docs, is `nonempty` supposed to be for `array` types? https://zod.dev/?id=nonempty I think you'll have to use `min()` here but you probably know more about this since you've been looking into the whole validating empty strings thing

File: packages/types/src/entities/ParticipantIdentification.ts [hunk 11]
```diff
@@ -41,10 +41,19 @@ export const hasRequiredGuardianInformation = (
...
+					guardianName: Name,
+					guardianPhoneNumber: PhoneNumber,
+					guardianRelationship: Name,
+				})
+				.safeParse({
+					guardianEmailAddress,
+					guardianName,
+					guardianPhoneNumber,
+					guardianRelationship,
+				}).success
```
Comment (alyssadsouza) [inline packages/types/src/entities/ParticipantIdentification.ts:56]: Shouldn't these already be validated from the schema? `.optional()` permits either `undefined` or the specified type, so checking that they're not `undefined` should mean they have the specified type right?

Comment (alyssadsouza) [inline packages/types/src/entities/ParticipantIdentification.ts:56]: Shouldn't these already be validated from the schema? `.optional()` permits either `undefined` or the specified type, so checking that they're not `undefined` should mean they have the specified type right?

File: apps/consent-ui/src/components/Form/fieldsets/CheckboxFieldSet/index.tsx [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+				</h4>
+			)}
+			<label htmlFor={name} className="checkbox-fieldset__label">
+				<CheckboxInput required={required} name={name} />
+				<span className="checkbox-fieldset__description">
+					{description}
+					{required && !title && <RequiredAsterisk />}
+				</span>
+			</label>
+			{error && <span style={{ color: 'red' }}>{error}</span>}
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/Form/fieldsets/CheckboxFieldSet/index.tsx:52]: I'm gonna do better styling in the form styling ticket, this is bad on purpose ;)

File: packages/types/src/entities/ClinicianInvite/ClinicianInviteForm.ts (hunk 10 above)
Comment (samrichca) [reply packages/types/src/entities/ClinicianInvite/ClinicianInviteForm.ts:38]: you're right, will change to min(1)

File: apps/consent-ui/src/components/Form/fieldsets/CheckboxFieldSet/index.tsx [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	name,
+	required = false,
+	title,
+}: FormCheckboxFieldSetProps<T>) => {
+	return (
+		<fieldset className={clsx('checkbox-fieldset', className)}>
+			{title && (
+				<h4>
+					{title}
+					{required && <RequiredAsterisk />}
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/Form/fieldsets/CheckboxFieldSet/index.tsx:42]: good point 👍

File: apps/consent-ui/src/i18n/locales/en/index.ts (hunk 4 above)
Comment (anncatton) [reply apps/consent-ui/src/i18n/locales/en/index.ts:48]: that was me, i also don't remember

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+	const [successMessageDemo, setSuccessMessageDemo] = useState('');
+	// TODO remove and redirect to homepage
+
+	const onSubmit: SubmitHandler<ClinicianInviteFormValidation> = (data, event) => {
+		event?.preventDefault();
+
+		const recaptchaToken = getRecaptchaToken();
+
+		if (recaptchaToken) {
+			axios
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:107]: any reason for limiting it to 200 originally? we'll need 201 & 204 which IIRC is allowed in axios by default

File: packages/types/src/entities/ParticipantIdentification.ts (hunk 11 above)
Comment (samrichca) [reply packages/types/src/entities/ParticipantIdentification.ts:56]: after thinking about this more, there's 3 potential outcomes

- undefined (skips non-`optional()` validation)
- defined but doesn't pass validation - short-circuits the validation, `optional()` isn't considered at all.
- defined & does pass validation

so the undefined check is fine and I'll revert to that. empty strings & whitespace are a separate issue that already has a ticket.

File: apps/consent-ui/src/i18n/locales/en/index.ts (hunk 4 above)
Comment (justincorrigible) [reply apps/consent-ui/src/i18n/locales/en/index.ts:48]: Not in the scope of this PR, but sounds to me like if there's no practical necessity for deviation from best practices (e.g. a weird third-party library requirement that breaks things otherwise), it'd be best if we return these to consistent JS casing.

Thoughts?

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (justincorrigible) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:107]: This could be abstracted into the response validation options within the Axios client.

That's done at `validateStatus` in the config object

[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:107]: removed validateStatus as discussed on slack

File: apps/consent-ui/src/components/Form/RequiredAsterisk/index.tsx [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import clsx from 'clsx';
+
+import styles from './RequiredAsterisk.module.scss';
+
+const RequiredAsterisk = ({ className }: { className?: string; required?: boolean }) => (
```
Comment (justincorrigible) [inline apps/consent-ui/src/components/Form/RequiredAsterisk/index.tsx:24]: the props `required` is unused and can be removed

File: apps/consent-ui/src/components/Form/fieldsets/inputs/SelectInput.tsx [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+	required,
+	value,
+}: FormSelectInputProps<T, V>) => (
+	<Select
+		aria-required={required}
+		instanceId={name}
+		name={name}
+		onChange={(val: FormSelectOnChangeArg<V>) =>
+			// the value could be a string or {label, value} object
+			onChange(typeof val === 'string' ? val : val?.value || null)
```
Comment (justincorrigible) [inline apps/consent-ui/src/components/Form/fieldsets/inputs/SelectInput.tsx:41]: 3 thoughts/questions:
1. cognitive burden: this line is quite loaded for this simple validation. It'd be easier to read (and possibly maintain) if abstracted and broken down as a switch.
2. consistency with types elsewhere: I was under the impression we wanted to use `undefined` instead of `null` for empty fields. Personally I'd prefer using nulls (difference between "field has been touched and left empty" vs "untouched field), but still, asking just to double-check for consistency.
3. consistency with types in the same file: as semi-related followup from the last point, in this line the fallback is `null`, while in line 45 the fallback is an empty string. There's likely a reason for this difference, but it isn't clear here. Should this line 41 also default to empty string instead?

File: apps/consent-ui/src/components/Form/fieldsets/FieldSet/index.tsx [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+	required = false,
+}: FormFieldSetProps<T>) => {
+	return (
+		<fieldset className={clsx('fieldset', className)}>
+			<label htmlFor={name}>
+				{label}
+				{required && <RequiredAsterisk />}
+			</label>
+			{children}
+			{error && <p style={{ color: 'red' }}>{error}</p>}
```
Comment (justincorrigible) [inline apps/consent-ui/src/components/Form/fieldsets/FieldSet/index.tsx:41]: Not asking for change, just sharing my own practices that I've found useful:
In layouts that include multiple different children/components, I tend to add blank lines to facilitate visually parsing the layout from source.

```suggestion
			</label>

			{children}

			{error && <p style={{ color: 'red' }}>{error}</p>}
```

File: apps/consent-ui/src/i18n/locales/en/index.ts (hunk 4 above)
Comment (anncatton) [reply apps/consent-ui/src/i18n/locales/en/index.ts:48]: ticket: https://github.com/OHCRN/platform/issues/289

**Review by anncatton — DISMISSED** [review]
Approving, but let's have @joneubank / @justincorrigible have final approval if possible

**Review by justincorrigible — DISMISSED** [review]
LGTM 👍 
Thanks for addressing my "consistency" nit picking, @samrichca.

Deferring to @joneubank for final approval

**Review by joneubank — CHANGES_REQUESTED** [review]
- Appreciate you putting so much detail into a docs file to help future devs.
- All the `trims` added to schemas is a good practice.

I feel like we can do better with creating the dictionaries for this form. I would like to see the use or Partial removed at the very least, and maybe a new function that returns an entire dictionary of a given language (rather than rebuilding it by requesting each property individually).

Other than that, I think it would be good to incorporate the type changes I recommended in the other PR.

File: packages/types/src/entities/Name.ts [hunk 18]
```diff
@@ -21,5 +21,5 @@ import { z } from 'zod';
 
 import { name } from './Regex.js';
 
-export const Name = z.string().regex(name);
+export const Name = z.string().trim().regex(name);
```
Comment (alyssadsouza) [inline packages/types/src/entities/Name.ts:24]: I was looking at the ticket you added where `Name` and `PhoneNumber` fail validation when the form returns no response as an empty string but 1) the regex doesn't accept an empty string, and 2) optional fields that don't have a value should be `undefined` according to how we treat them in all other schemas #285 

Maybe we could update the regex to allow an empty string, and add a `NameOptional` schema that does a `transform` on empty strings to convert them to `undefined`, like this:

```ts
const name = /^[A-Za-z\s]*$/ // using `*` instead of `+` so it allows 0 occurrences

export const NameOptional = z
	.string()
	.trim()
	.regex(name)
	.transform((x) => x || undefined);
```

Then for the original `Name` schema just add a `min(1)` constraint:
```ts
export const Name = z.string().trim().regex(name).min(1);
```

This could go in a separate PR for that ticket, just leaving it here in case you have any thoughts

File: apps/consent-ui/src/components/ClinicianInviteForm/types.ts [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+import { FormsDictionary } from 'src/i18n/locales/en/forms';
+import { ConsentGroupDictionary } from 'src/i18n/locales/en/consent-group';
+
+import { FormSelectOption } from '../Form/types';
+
+// TODO ClinicianInviteFormDictionary shouldn't be partial in final version
+// i.e. when styling/page is complete.
+export type ClinicianInviteFormTextDictionary = Partial<
+	ClinicianInviteFormDictionary & FormsDictionary & FormErrorsDictionary
+>;
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/types.ts:31]: not finished adding everything to the page yet, but this will be updated by me shortly.

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+	} = useRecaptcha();
+
+	const handleRecaptchaChange = () => {
+		const token = getRecaptchaToken();
+		token && setRecaptchaError('');
+		onRecaptchaChange();
+	};
+
+	// submit form
+	const [successMessageDemo, setSuccessMessageDemo] = useState('');
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx:95]: it can be removed, we can check the network tab to see if the request worked.

File: packages/types/src/entities/Name.ts (hunk 18 above)
Comment (samrichca) [reply packages/types/src/entities/Name.ts:24]: thanks alyssa! added to appropriate ticket.

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+
+				{successMessageDemo && (
+					<Notification level="success" variant="small" title={successMessageDemo} />
+				)}
+			</Form>
+		</FormProvider>
+	);
+};
+
+export default ClinicianInviteFormComponent;
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: moved my files to components/views & components/common and made an additional ticket

File: apps/consent-ui/src/app/[lang]/invite/index.tsx [hunk 22]
```diff
@@ -18,15 +18,73 @@
...
+import { FormErrorsDictionary } from 'src/i18n/locales/en/form-errors';
 
 const ClinicianRegistration = async ({ currentLang }: { currentLang: ValidLanguage }) => {
 	const translate = getTranslation(currentLang);
+
+	const errorsDict: FormErrorsDictionary = {
+		required: translate('form-errors', 'required'),
+	};
+
+	const labelsDict: Partial<FormLabelsDictionary> = {
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/app/[lang]/invite/index.tsx:39]: I ran into issues implementing this feedback with the current `kebab-case` in translation dictionaries. I made a ticket for changing kebab to camel, and a follow-up ticket for this form https://github.com/OHCRN/platform/issues/300

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+
+const ClinicianInviteFormComponent = ({
+	consentGroupOptions,
+	errorsDict,
+	labelsDict,
+	textDict,
+}: {
+	consentGroupOptions: ConsentGroupOption[];
+	errorsDict: FormErrorsDictionary;
+	labelsDict: Partial<FormLabelsDictionary>;
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: I ran into issues implementing this feedback with the current `kebab-case` in translation dictionaries. I made a ticket for changing kebab to camel, and a follow-up ticket for this form https://github.com/OHCRN/platform/issues/300

[the remark this answers is not in the document]
Comment (anncatton) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: what was the issue you ran into? i'm confused how the key case would affect creating types

[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: I wanted to deconstruct the translations 

`const { 'kebab-case' } = ...`

[the remark this answers is not in the document]
Comment (justincorrigible) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: Right, I've been thinking about this one, and as a quick stopgap to unblock you without ballooning the scope, you could destructure the ones you can, and separately have things like:
`const camelCase = dictionary['kebab-case'];`

which you could then remove on a cleanup followup PR that also fixes the casing.

File: apps/consent-ui/src/app/[lang]/invite/index.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (anncatton) [reply apps/consent-ui/src/app/[lang]/invite/index.tsx:39]: A followup question to what Jon mentioned - just to make sure I understand the goal with the new ticket: it's to make dictionaries for *each* form as we build them, correct? i.e. a ClinicianInviteFormDict, a ConsentWizardStep2Dict, etc etc? That was my original intention with having dictionaries dedicated to particular components, as opposed to trying to over-generalize with a "FormDictionary" of sorts.

Because even here, the `labelsDict` works with `satisfies FormLabelsDictionary` without the use of `Partial`.

File: apps/consent-ui/src/components/common/forms/fieldsets/inputs/SelectInput.tsx [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+const SelectInput = <T extends FieldValues, V extends string>({
+	name,
+	onChange,
+	options,
+	placeholder = '',
+	required,
+	value,
+}: FormSelectInputProps<T, V>) => {
+	return (
+		<Select
```
Comment (mistryrn) [inline apps/consent-ui/src/components/common/forms/fieldsets/inputs/SelectInput.tsx:36]: Not sure if this is a limitation of the `react-select` library, but I can't seem to find where/how it applies `label`s to the select `input`s.

When running the [aXe DevTools](https://chromewebstore.google.com/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd) I get a Critical Issue relating to the select input created, stating the form element does not have a label:
Screen Shot 2023-12-06 at 4 13 08 PM

Can you look into this and see if there's some way to have it apply an `aria-label` of the `value`, or something like that (or a better solution than that :P)

File: apps/consent-ui/src/components/views/invite/ClinicianInviteForm/index.tsx [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+import RecaptchaCheckbox from '../../../common/forms/RecaptchaCheckbox';
+
+import { ClinicianInviteFormTextDictionary, ConsentGroupOption } from './types';
+
+const consentGroupsRequiringGuardian: ConsentGroup[] = [
+	ConsentGroup.enum.GUARDIAN_CONSENT_OF_MINOR,
+	ConsentGroup.enum.GUARDIAN_CONSENT_OF_MINOR_INCLUDING_ASSENT,
+];
+
+const guardianInfoFields: Partial<keyof ClinicianInviteRequest>[] = [
```
Comment (anncatton) [inline apps/consent-ui/src/components/views/invite/ClinicianInviteForm/index.tsx:48]: Is it possible to reuse the key names from `InviteGuardianFields` here instead?

**Review by mistryrn — COMMENTED** [review]
Lots of good stuff in here, I really like how modular the form components are! Just a few comments/questions below.

I checked out the branch and built/ran it locally, then did a quick aXe DevTools scan – we might need to take a look at how that `react-select` library works, since it's currently giving us a Critical a11y issue related to input labels (screenshot provided on PR comment).

One of the translations isn't currently showing up for me (`textDict['clinician-information']`), and I wasn't sure about the use of `{ required: true }` in the `rules` prop for the `SelectFieldSet`'s `Controller`.

Other than that, the only major question I had was regarding other errors, for things other than just `required` – I don't see that handled in this PR, but not sure if that was part of the scope or if it has already been discussed.

File: apps/consent-ui/src/components/ClinicianInviteForm/index.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (anncatton) [reply apps/consent-ui/src/components/ClinicianInviteForm/index.tsx]: seems like in that case it would be better - if possible - to add this deconstruction in after the case change pr is in...i think you already have that work on the go?

File: apps/consent-ui/src/components/common/forms/fieldsets/inputs/SelectInput.tsx (hunk 24 above)
Comment (samrichca) [reply apps/consent-ui/src/components/common/forms/fieldsets/inputs/SelectInput.tsx:36]: I had to change instanceId to inputId, my b.

File: apps/consent-ui/src/components/views/invite/ClinicianInviteForm/index.tsx (hunk 25 above)
Comment (samrichca) [reply apps/consent-ui/src/components/views/invite/ClinicianInviteForm/index.tsx:48]: I can't reply directly to your other comment

I'd like to have dictionaries for each form/component but I was thinking we could reuse some of the form labels and have one source of truth for translating them, e.g. we can get "first name/prénom" from a shared dictionary. is that alright or should they be completely separate?

File: apps/consent-ui/src/app/[lang]/invite/index.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/app/[lang]/invite/index.tsx]: yes, I want to have separate dictionaries for all the forms for type safety, but I'm hoping they can share some of the translations from `FormLabelsDictionary`

File: apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+	// watch consentGroup value & show/hide guardian info fields if participant is a minor.
+	const watchConsentGroup = watch('consentGroup');
+	const [showGuardianFields, setShowGuardianFields] = useState<boolean>(false);
+	useEffect(() => {
+		if (consentGroupsRequiringGuardian.includes(watchConsentGroup)) {
+			// guardian fields are registered on render, in their input components
+			setShowGuardianFields(true);
+		} else {
+			setShowGuardianFields(false);
+			guardianInfoFields.forEach((field: Partial<keyof ClinicianInviteRequest>) => {
```
Comment (anncatton) [inline apps/consent-ui/src/components/views/Invite/ClinicianInviteForm/index.tsx:124]: i think the type on `field` can be removed since `guardianInfoFields` is already typed above

**Review by anncatton — COMMENTED** [review]
Thanks for all the comments/TODO ticket links @samrichca ! I just added one comment to add one more TODO ticket link, and a minor type change.

I'd like @mistryrn and @justincorrigible to leave a final approval for any changes they requested; i think the latest changes + the followup tickets address the rest of @joneubank 's feedback but i'd like their 👀 there too for a sanity check.

Can we link the followup tickets to the original #186 too?

File: apps/consent-ui/src/components/common/Form/types.ts [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+export type FormSelectOnChangeArg<V extends string> = SingleValue<string | FormSelectOption<V>>;
+
+// setup types for react-hook-forms API
+type FormFieldName<T extends FieldValues> = Path<T>;
+
+// fieldsets that use the FieldSet component
+
+interface FormFieldSetSharedProps<T extends FieldValues> {
+	className?: string;
+	error?: any; // TODO harmonize error dictionary & RHF errors
```
[the remark this answers is not in the document]
Comment (samrichca) [reply apps/consent-ui/src/components/common/Form/types.ts:41]: added

**Comment by samrichca** [comment]
added follow up tickets to this ticket

**Review by mistryrn — COMMENTED** [review]
All the recent changes look good! Thanks for highlighting the follow-up tickets where applicable 🙏 

My only outstanding question/comment (apologies if this has already been covered earlier in the PR process) is regarding the optional "Preferred Name" field: have we figured out how to make it actually optional on the form? I'm still getting a "required" message when I try to submit the form without filling out that field:
Screen Shot 2023-12-13 at 1 23 00 PM

**Comment by samrichca** [comment]
@mistryrn very good question, there's another followup ticket for that. https://github.com/OHCRN/platform/issues/285

**Review by mistryrn — APPROVED** [review]
Thanks for clarifying/listing out all the follow-up tickets, @samrichca !

Since the follow-up/out-of-scope work is captured in the tickets you listed in this comment (https://github.com/OHCRN/platform/issues/186#issuecomment-1854194859), I'm good to proceed with merging this PR.

Do we have a plan/order/priority for those follow-up tasks?

I think we'll still need @joneubank to give final approval since his prior change request will block you from merging though.

**Comment by samrichca** [comment]
thanks @mistryrn, I've updated that comment with details about priority and status.

if @joneubank isn't available tomorrow morning I'll see if @justincorrigible can merge it

**Review by anncatton — APPROVED** [review]
